pandas/_libs/lib.pyx

```diff
@@ -1414,10 +1414,12 @@ def infer_dtype(value: object, skipna: bool = True) -> str:
             return "time"
 
     elif is_decimal(val):
-        return "decimal"
+        if all(is_decimal(x) for x in values):
```

Comment:
can you create is_decimal_array to have the same style

Reply:
i think you actually want to use the validator e.g as a raw NaN would not work here

Reply:
can you expand on this?  do you expect something other than Decimal objects to work here?

Reply:
NaNs are valid as well here; but more important is that every validator has the same pattern. a is_* function e.g. is_decimal_array, which should call here.

Reply:
updated to use the is_* pattern.

since decimal has Decimal("nan"), i wouldnt expect to accept np.nan here

Reply:
> since decimal has Decimal("nan"), i wouldnt expect to accept np.nan here

yeah ok, though i think we could actually accept np.nan as well and infer a decimal if that is the case (as we do this for all other dtypes). can you open an issue.